azext_edge/edge/providers/orchestration/upgrade2.py

```diff
@@ -0,0 +1,309 @@
+
+        # TODO @digimaun - deterine further pre-checks.
+        if not self.extensions_map.get(EXTENSION_TYPE_OPS):
+            raise ValidationError(
+                "The cluster backing the instance has an invalid state. IoT Operations extension not detected."
+            )
+
+        for ext_type in EXTENSION_TYPE_TO_MONIKER_MAP:
+            ext_moniker = EXTENSION_TYPE_TO_MONIKER_MAP[ext_type]
+            extension = self.extensions_map.get(ext_type)
+            if extension:
+                ext_queue.append(
+                    ExtensionUpgradeState(
+                        extension=extension,
+                        desired_version_map=self.init_version_map.get(ext_moniker, {}),
+                        override=self.override_map.get(ext_moniker),
+                    )
+                )
+        return ext_queue
+
+
+class ExtensionUpgradeState:
+    def __init__(self, extension: dict, desired_version_map: dict, override: Optional[ConfigOverride] = None):
+        self.extension = extension
+        self.desired_version_map = desired_version_map
+        self.override = override or ConfigOverride()
+
+    @property
+    def current_version(self) -> Tuple[str, str]:
+        return (self.extension["properties"]["version"], self.extension["properties"]["releaseTrain"])
+
+    @property
+    def desired_version(self) -> Tuple[str, str]:
+        return (
+            self.override.version or self.desired_version_map.get("version"),
+            self.override.train or self.desired_version_map.get("train"),
+        )
+
+    @property
+    def moniker(self) -> str:
+        return EXTENSION_TYPE_TO_MONIKER_MAP[self.extension["properties"]["extensionType"].lower()]
+
+    def can_upgrade(self) -> bool:
+        return any(
+            [
+                self._has_delta_in_version(),
+                self._has_delta_in_train(),
+                self._has_delta_in_config(),
+            ]
+        )
+
+    def get_patch(self) -> dict:
+        if not self.can_upgrade():
+            return {}
+
+        payload = {
+            "properties": {},
+        }
+
+        if self._has_delta_in_version():
+            payload["properties"]["version"] = self.desired_version[0]
+        if self._has_delta_in_train():
+            payload["properties"]["releaseTrain"] = self.desired_version[1]
+        if self._has_delta_in_config():
+            payload["properties"]["configurationSettings"] = self.override.config
+
+        return payload
+
+    def _has_delta_in_version(self) -> bool:
+        return bool(self.override.version) or (
+            self.desired_version[0] and version.parse(self.desired_version[0]) > version.parse(self.current_version[0])
+        )
+
+    def _has_delta_in_train(self) -> bool:
+        return bool(self.override.train) or (
+            self.desired_version[1] and self.desired_version[1].lower() != self.current_version[1].lower()
+        )
+
+    def _has_delta_in_config(self) -> bool:
```

Comment:
are we going to check the desired config states?

Reply:
You are talking about some delta between extension X 1.0 -> X 2.0 right? Lets cross that bridge when we get there.